In [6]:
#!pip install ast

In [41]:
import pandas as pd
import numpy as np
import ast

In [42]:
content_df = pd.read_csv("/Users/shriyashekhar/Documents/IWork/Data/Output/content_nouns.csv", encoding="ISO-8859-1")
content_df.head()

StreamID                                            Content  \
0       163  TXmAk2KZAy4\r\nNMeUjebo1Ac\r\nEEuTxFhp3go\r\nC...   
1       419  TXmAk2KZAy4\r\nNMeUjebo1Ac\r\nEEuTxFhp3go\r\nC...   
2       507  wBYKUgUyGWc\r\nA team of world-class drivers, ...   
3       199  Castrol EDGE  is Castrol?s flagship power bran...   
4       201  Charles ?Cheers? Wakefield, Castrol?s founder,...   

                                       Content_Nouns  
0  [('Castrol', 'NNP'), ('oil', 'NN'), ('TITANIUM...  
1  [('Castrol', 'NNP'), ('oil', 'NN'), ('TITANIUM...  
2  [('wBYKUgUyGWc', 'NN'), ('team', 'NN'), ('worl...  
3  [('Castrol', 'NNP'), ('EDGE', 'NNP'), ('Castro...  
4  [('Charles', 'NNP'), ('\x93Cheers\x94', 'NNP')...

In [43]:
def get_nouns_for_stream(stream_id):
    stream_nouns = content_df[content_df["StreamID"] == stream_id]["Content_Nouns"].values
    #stream_nouns = stream_nouns.astype(str)
    first_stream_noun = stream_nouns[0]
#     print("Printing debug info....")
#     print(first_stream_noun)
#     print(type(first_stream_noun))
#     print("End of printing debug info....")
    
    if isinstance(first_stream_noun, str):
        # https://stackoverflow.com/questions/10775894/converting-a-string-representation-of-a-list-into-an-actual-list-object
        first_stream_noun = ast.literal_eval(first_stream_noun)
        
    return [x[0] for x in first_stream_noun]

### Nouns for the stream 507 as a list as below...

In [44]:
get_nouns_for_stream(199)

['Castrol',
 'EDGE',
 'Castrol\x92s',
 'flagship',
 'power',
 'brand',
 'PCOs',
 'performance',
 'engine',
 'oil',
 'performing',
 'vehicles',
 'OEMs',
 'world',
 'Castrol',
 'EDGE',
 'TITANIUM',
 'FST\x99',
 'oil',
 'strength',
 'oil',
 'film',
 'protection',
 'oil',
 'pressures',
 'engines',
 'today',
 'video',
 'Today',
 'engines',
 'pressure',
 'areas',
 'engine',
 'pressures',
 'excess',
 'square',
 'centimetre',
 'that\x92s',
 'cars',
 'stiletto',
 'hee',
 'video',
 'Castrol',
 'EDGE',
 'strength',
 'technology',
 'pressures',
 'high-performance',
 'engines',
 'Castrol',
 'EDGE',
 'problem',
 'engine',
 'pressures',
 'TITANIUM',
 'FST\x99',
 'Fluid',
 'Strength',
 'Technology',
 'TITANIUM',
 'FST\x99',
 'oil',
 'strength',
 'oil',
 'breakdown',
 'sha',
 'TITANIUM',
 'FST\x99',
 'adapts',
 'engine',
 'need',
 'pressures',
 'contact',
 'variety',
 'engine',
 'spee',
 'Castrol',
 'R',
 'amp',
 'D',
 'Technology',
 'world',
 'Castrol',
 'Stress',
 'Test',
 'film',
 'strength',
 'Cast

## Shriya: So you call the method like this: get_nouns_for_stream(507) to get a list of nouns for that stream with the specified stream ID. You can then loop through the items in the list and do your word net and SPARQL analysis.

In [45]:
from nltk.corpus import wordnet as wn

list_of_nouns = get_nouns_for_stream(201)
print(len(list_of_nouns))

for i in range(0, len(list_of_nouns)):
    for j in range(0, len(list_of_nouns)):
# for noun in list_of_nouns:
#     print(i)
#     if not wn.synsets(list_of_nouns[i]):
        while(len(wn.synsets(list_of_nouns[i]))==0):
            i+=1
        a = wn.synsets(list_of_nouns[i])[0]
#         i+=1
        while(len(wn.synsets(list_of_nouns[j]))==0):
    #         print(i)
            j+=1
        print(j)
        b = wn.synsets(list_of_nouns[j])[0]
        if a and b:
            print('two terms')
            print(a)
            print(b)
            print('parent')
            print(a.lowest_common_hypernyms(b))


# return a.lowest_common_hypernyms(b)

100
0
two terms
Synset('charles.n.01')
Synset('charles.n.01')
parent
[Synset('charles.n.01')]
4
two terms
Synset('charles.n.01')
Synset('laminitis.n.01')
parent
[Synset('entity.n.01')]
4
two terms
Synset('charles.n.01')
Synset('laminitis.n.01')
parent
[Synset('entity.n.01')]
4
two terms
Synset('charles.n.01')
Synset('laminitis.n.01')
parent
[Synset('entity.n.01')]
4
two terms
Synset('charles.n.01')
Synset('laminitis.n.01')
parent
[Synset('entity.n.01')]
5
two terms
Synset('charles.n.01')
Synset('entrepreneur.n.01')
parent
[Synset('person.n.01')]
6
two terms
Synset('charles.n.01')
Synset('vision.n.01')
parent
[Synset('entity.n.01')]
7
two terms
Synset('charles.n.01')
Synset('lubricant.n.01')
parent
[Synset('physical_entity.n.01')]
8
two terms
Synset('charles.n.01')
Synset('business.n.01')
parent
[Synset('entity.n.01')]
10
two terms
Synset('charles.n.01')
Synset('london.n.01')
parent
[Synset('object.n.01')]
10
two terms
Synset('charles.n.01')
Synset('london.n.01')
parent
[Synset('object.

IndexError: list index out of range

In [46]:
from nltk.corpus import wordnet as wn
words = get_nouns_for_stream(201)
syns = {w : [] for w in words}

for k, v in syns.items():
    for synset in wn.synsets(k):
        for lemma in synset.lemmas():
            v.append(lemma.name())
print(syns)

{'Charles': ['Charles', 'Charles_IX', 'Charles', 'Charles_VII', 'Charles', 'Charles_II', 'Charles_I', 'Charles_the_Bald', 'Charles', 'Charles_II', 'Charles', 'Charles_I', 'Charles_Stuart', 'Charles', 'Prince_Charles', 'Charles', 'Jacques_Charles', 'Jacques_Alexandre_Cesar_Charles', 'Charlemagne', 'Carolus', 'Charles', 'Charles_I', 'Charles_the_Great', 'Charles', 'Charles_River'], '\x93Cheers\x94': [], 'Wakefield': [], 'Castrol\x92s': [], 'founder': ['laminitis', 'founder', 'founder', 'beginner', 'founding_father', 'father', 'founder', 'fall_through', 'fall_flat', 'founder', 'flop', 'founder', 'collapse', 'fall_in', 'cave_in', 'give', 'give_way', 'break', 'founder', 'founder'], 'entrepreneur': ['entrepreneur', 'enterpriser'], 'vision': ['vision', 'sight', 'vision', 'visual_sense', 'visual_modality', 'vision', 'visual_sensation', 'imagination', 'imaginativeness', 'vision', 'vision'], 'lubricant': ['lubricant', 'lubricator', 'lubricating_substance', 'lube'], 'business': ['business', 'conc

In [47]:
dog = wn.synsets('dog')[0]
cat = wn.synsets('cat')[0]
cat.lowest_common_hypernyms(dog)

[Synset('carnivore.n.01')]

In [48]:
oil = wn.synsets('oil')[1]
castrol = wn.synsets('liquid')[0]
oil.lowest_common_hypernyms(castrol)

[Synset('substance.n.01')]

In [49]:
wn.synset('policeman.n.01').lowest_common_hypernyms(wn.synset('chef.n.01'))

[Synset('person.n.01')]

In [50]:
bata = wn.synsets('Wakefield')
if bata:
    print("empty")

In [51]:
len(wn.synsets('liquid'))

11

In [52]:
wn.synsets('liquid')[0]

Synset('liquid.n.01')

In [53]:
import nltk
from nltk.corpus import wordnet as wn
dog = wn.synset('charles.n.01')
cat = wn.synset('power.n.01')
float(dog.path_similarity(cat))

0.07692307692307693

In [54]:
from nltk.corpus import wordnet as wn

list_of_nouns_1 = get_nouns_for_stream(201)
list_of_nouns_2 = get_nouns_for_stream(199)
print(list_of_nouns_1)
a_1 = []
b_1 = []
for i in range(0, len(list_of_nouns_1)-1):
    for j in range(0, len(list_of_nouns_2)-1):
# for noun in list_of_nouns:
#     print(i)
#     if not wn.synsets(list_of_nouns[i]):
        while(len(wn.synsets(list_of_nouns_1[i]))==0):
            i+=1
            a_1.append(list_of_nouns_1[i])
        a = wn.synsets(list_of_nouns_1[i])[0]
#         i+=1
        while(len(wn.synsets(list_of_nouns_2[j]))==0):
    #         print(i)
            j+=1
            b_1.append(list_of_nouns_2[j])
#         print(j)
        b = wn.synsets(list_of_nouns_2[j])[0]
        if a and b:
#             print('two terms')
#             print(a)
#             print(b)
#             print('parent')
#             print(a.lowest_common_hypernyms(b))
            x = a.path_similarity(b)
            if((x is not None) and (x >= 0.5) and (a is not b)):
                
#                 print('similar terms')
#                 print(a)
#                 print(b)
#                 print('hypernym')
#                 print(a.lowest_common_hypernyms(b))
                a_1.append(a.lowest_common_hypernyms(b))
                b_1.append(a.lowest_common_hypernyms(b))
#                 print(a_1)
print(a_1)

# return a.lowest_common_hypernyms(b)

['Charles', '\x93Cheers\x94', 'Wakefield', 'Castrol\x92s', 'founder', 'entrepreneur', 'vision', 'lubricant', 'business', 'Cheapside', 'London', 'Castrol', 'interest', 'contrapt', 'Castrol', 'NASA', 'lubricants', 'space', 'shuttle', 'programme', 'NASA', 'Curiosity', 'rover', 'mission', 'Mars', 'Castrol', 'grease', 'role', 'opera', 'journey', 'oil', 'car', 'engine', 'secs.In', 'time', 'engine', 'parts', 'prevents', 'film', 'journey', 'Castrol', 'seperate', 'brand', 'heritage', 'personality', 'Castrol', 'flight', 'Castrol', 'role', 'NASA', 'Curiosity', 'rover', 'Mars', 'operation', 'Curiosity', 'wheels', 'Charles', '\x93Cheers\x94', 'Wakefield', 'Castrol\x92s', 'founder', 'entrepreneur', 'vision', 'lubricant', 'business', 'Cheapside', 'London', 'interest', 'contraptions', 'automobile', 'th', 'Having', 'kind', 'motor', 'oil', 'CC', 'Wakefield', 'method', 'customers', 'product', 'sponsorship', 'Castrol', 'name', 'banners', 'flags', 'aviation', 'events', 'British', 'Petroleum', 'BP', 'Castro